app/services/appointment_scheduler.rb

```diff
@@ -1,48 +1,115 @@
-                                      ubs_id: ubs.id, patient_id: nil)
-        .order('random()') # reduces row lock contention
-        .first
+  # Scheduler for appointments, enabling appointments to be made between +earliest_allowed+ and +latest_allowed+ dates
+  def initialize(earliest_allowed:, latest_allowed:)
+    @earliest_allowed = earliest_allowed
+    @latest_allowed = latest_allowed
+  end
 
-      return [:all_slots_taken] unless appointment.present?
+  # Schedules an appointment for +patient+ starting +from+ a given timestamp, optionally with a +ubs_id+ (which can be
+  # nil in case we want to schedule on any Ubs available to the +patient+).
+  # It will schedule on the first possible time until the +latest_allowed+.
+  # Checks if the user can schedule, otherwise returns +CONDITIONS_UNMET+.
+  # Looks for appointments, and tries to schedule one. If it can't, it will return +NO_SLOTS+.
+  # In case it can, it will also cancel the patient's current schedule for an existing appointment, and in the end it
+  # returns +SUCCESS+ and the newly scheduled appointment.
+  def schedule(patient:, ubs_id:, from:)
+    return [CONDITIONS_UNMET] unless patient.can_schedule?
 
-      patient.appointments.future.update_all(patient_id: nil)
+    current_appointment = patient.appointments.current
 
-      appointment.update!(
+    Appointment.transaction(isolation: :repeatable_read) do
+      success = update_appointment(
         patient_id: patient.id,
-        second_dose: patient.appointments.current&.check_out.present?,
-        vaccine_name: patient.appointments.current&.vaccine_name
+        ubs_id: ubs_id,
+        start: rounded_from(from)..latest_allowed
       )
+      return [NO_SLOTS] unless success
 
-      # TODO: remove this after we get rid of last_appointment
-      patient.update!(last_appointment: appointment)
+      new_appointment = patient.appointments.open.where.not(id: current_appointment&.id).first!
+      if current_appointment
+        # Update new appointment with data from current
+        new_appointment.update!(
+          vaccine_name: current_appointment.vaccine_name
+        )
 
-      [:success, appointment]
+        # Free up current appointment if it wasn't checkout out (completed)
+        unless current_appointment.checked_out?
+          current_appointment.update!(patient: nil, check_in: nil, check_out: nil, second_dose: nil, vaccine_name: nil)
+        end
+      end
+
+      [SUCCESS, new_appointment]
     end
-  rescue ActiveRecord::SerializationFailure
-    [:all_slots_taken]
-  rescue => e
-    Sentry.capture_exception(e)
+  end
+
+  # Cancels schedule for an appointment for a given user, in a SQL efficient way
+  def cancel_schedule(patient:, id:)
+    patient.appointments
+           .open
+           .where(id: id)
+           .update_all(patient_id: nil, updated_at: Time.zone.now) # rubocop:disable Rails/SkipsModelValidations
```

Comment:
Why `update_all` if it's scoped by `id`?

Reply:
I think it's the only way to have an actual `UPDATE` statement (except for writing manual SQL, of course). `.find(id).update(...)` would result in two queries.

Reply:
Exactly, it's to make it as efficient as possible, hence why the manual update_all.